src/lib/Microsoft.Health.Fhir.Ingest/Data/MeasurementToEventMessageAsyncCollector.cs

```diff
@@ -54,51 +53,49 @@ public async Task AddAsync(IEnumerable<IMeasurement> items, CancellationToken ca
 
             using (var hasher = _hashCodeFactory.CreateDeterministicHashCodeGenerator())
             {
-                var submissionTasks = items
-                .GroupBy(m =>
-                {
-                    // cast as byte to restrict to 256 possible values. This will lead to a greater change of measurements ending up in the same bucket,
-                    // while providing partition keys with enough entropy for EventHub to better distribute them across partitions.
-                    return hasher.GenerateHashCode(m.DeviceId.ToLower());
-                })
-                .Select(async grp =>
+                var batches = new Dictionary<string, EventDataBatch>();
+                var submissionTasks = new List<Task>();
+
+                foreach (var m in items)
                 {
-                    var partitionKey = grp.Key;
-                    var currentEventDataBatch = await _eventHubService.CreateEventDataBatchAsync(partitionKey);
+                    var partitionKey = hasher.GenerateHashCode(m.DeviceId.ToLower());
+                    EventDataBatch eventDataBatch = null;
+                    if (!batches.TryGetValue(partitionKey, out eventDataBatch))
+                    {
+                        eventDataBatch = await _eventHubService.CreateEventDataBatchAsync(partitionKey);
+                        batches.Add(partitionKey, eventDataBatch);
+                    }
 
-                    foreach (var m in grp)
+                    var currentEventDataInBytes = System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(m);
+                    var eventData = new EventData(currentEventDataInBytes);
+
+                    if ((eventDataBatch.SizeInBytes >= _bufferSizeLimit) || !eventDataBatch.TryAdd(eventData))
```

Comment:
The big change here was sending small chunks of the Measurements across multiple tasks.